molecule 3.4.0 (rev 1) - add docker driver

mpiekunk opened the pull request:
Molecule uses Drivers to manage instances to operate on. This adds the
Docker driver which is widely used to quickly test against multiple OS
versions. The Vagrant driver was already added.

Using the Vagrant & Docker drivers are optional, but may require additional
dependencies to be installed. Added the caveats section to note these
requirements which weren't captured before.

Added test to validate that Docker and Vagrant drivers were installed.
Changed existing test to validate default Molecule functionality instead
of a specific driver (Vagrant) since driver use is optional.

- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/HEAD/CONTRIBUTING.md)?
- [x] Have you ensured that your commits follow the [commit style guide](https://docs.brew.sh/Formula-Cookbook#commit)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Is your test running fine `brew test <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install --build-from-source <formula>`)? If this is a new formula, does it pass `brew audit --new <formula>`?

-----

carlocab commented on Formula/molecule.rb:
```diff
@@ -19,7 +20,6 @@ class Molecule < Formula
   depends_on "ansible"
   depends_on "openssl@1.1"
   depends_on "python@3.9"
-  depends_on "six"
```
Why change this?

mpiekunk replied:
I had a question about this one.  I had a problem with Molecule initially where i updated the Molecule package but not the Six package.  This caused an error due to Molecule needing a newer version.  Since Six is a library and not a tool/utility you run, it seemed to fit better as a versioned resource in the formula instead of a brew dependency.  This would also remove any version mismatch issues in the future.  Thoughts on which is best?  Wasn't confident what the preferred method was after looking at the [Python for Formula Authors](https://docs.brew.sh/Python-for-Formula-Authors) docs.

carlocab replied:
Molecule has a dependency on Six, so upgrading Molecule should also upgrade Six.

mpiekunk replied:
Sorry.... I was remembering incorrectly what the issue I had was.  During the installation of Molecule I got this error:

```
==> Installing molecule dependency: six
==> Pouring six--1.16.0_1.catalina.bottle.tar.gz
Error: The `brew link` step did not complete successfully
The formula built, but is not symlinked into /usr/local
Could not symlink lib/python3.9/site-packages/six.py
Target /usr/local/lib/python3.9/site-packages/six.py
already exists. You may want to remove it:
  rm '/usr/local/lib/python3.9/site-packages/six.py'

To force the link and overwrite all conflicting files:
  brew link --overwrite six

To list all files that would be deleted:
  brew link --overwrite --dry-run six

Possible conflicting files are:
/usr/local/lib/python3.9/site-packages/six.py
```

That's why I moved it from a dependency to resource block.  My question about which approach is better still applies.

carlocab replied:
Yea, that happens when you `pip install` into Homebrew's `site-packages` directory, and not just for `six`. I suggest you do `pip install -U` or install into a virtual env.

Let's keep the `six` dependency -- it avoids duplicating code across different formulae, and duplicating installs of the same Python module on the same machine.

carlocab commented on Formula/molecule.rb:
```diff
@@ -104,23 +109,18 @@ class Molecule < Formula
     sha256 "83f5e5a09f9c5f68f60173de572930effbcc0287bb84fdc4426cb4168c088424"
   end
 
-  resource "docker-py" do
-    url "https://files.pythonhosted.org/packages/fa/2d/906afc44a833901fc6fed1a89c228e5c88fbfc6bd2f3d2f0497fdfb9c525/docker-py-1.10.6.tar.gz"
-    sha256 "4c2a75875764d38d67f87bc7d03f7443a3895704efc57962bdf6500b8d4bc415"
-  end
```
If we're removing this we need to also remove it from `pypi_formula_mappings.json`.

mpiekunk replied:
I wasn't aware `pypi_formula_mappings.json` needed to be updated, but I will do that.  When adding new resources do they need to be added to the mappings as well?  Finally, is this something that should be added to the [Python for Formula Authors](https://docs.brew.sh/Python-for-Formula-Authors) docs or is it documented somewhere else and just needs to be linked?

carlocab replied:
It depends. If they're already found as default dependencies by `pipgrip`, then it's not needed. If these are additional Python modules that aren't found via `pipgrip`, they need to be added there.

carlocab commented on Formula/molecule.rb:
```diff
@@ -203,10 +208,10 @@ class Molecule < Formula
-    url "https://files.pythonhosted.org/packages/bb/c6/0a6d22ae1782f261fc4274ea9385b85bf792129d7126575ec2a71d8aea18/python-vagrant-0.5.15.tar.gz"
-    sha256 "af9a8a9802d382d45dbea96aa3cfbe77c6e6ad65b3fe7b7c799d41ab988179c6"
-  end
+  # resource "python-vagrant" do
+  #   url "https://files.pythonhosted.org/packages/bb/c6/0a6d22ae1782f261fc4274ea9385b85bf792129d7126575ec2a71d8aea18/python-vagrant-0.5.15.tar.gz"
+  #   sha256 "af9a8a9802d382d45dbea96aa3cfbe77c6e6ad65b3fe7b7c799d41ab988179c6"
+  # end
```
Either remove it or leave it uncommented, please.

mpiekunk replied:
Sorry.  This was an oversight and I will fix it.

carlocab commented on Formula/molecule.rb:
```diff
@@ -238,6 +243,11 @@ class Molecule < Formula
     sha256 "d435f514e834e3fdc0941f6a29d086b80b2ea51b28112aee6254bd104ee42a74"
   end
 
+  resource "six" do
+    url "https://files.pythonhosted.org/packages/71/39/171f1c67cd00715f190ba0b100d606d440a28c93c7714febeca8b79af85e/six-1.16.0.tar.gz"
+    sha256 "1e61c37477a1626458e36f7b1d82aa5c9b094fa4802892072e49de9c60c4c926"
+  end
+
```
```suggestion
```
We don't need this if you keep `depends_on "six"`.

carlocab commented on Formula/molecule.rb:
```diff
@@ -264,23 +274,43 @@ class Molecule < Formula
     virtualenv_install_with_resources
   end
 
+  def caveats
+    <<~EOS
+
+      Depending on the Molecule driver chosen, you may need to install additional packages.
+
+      Molecule-Docker Driver
+      ----------------------
+        - Docker
+        - community.docker Ansible collection
+
+      Molecule-Vagrant Driver
+      -----------------------
+        - VirtualBox
+        - Vagrant
+
+    EOS
+  end
```
Is this not in upstream documentation?

mpiekunk replied:
The general comment about "you may need additional packages depending on the driver" is mentioned in the [Molecule Docs](https://molecule.readthedocs.io/en/latest/installation.html#requirements) but specific requirements for the Vagrant & Docker drivers are scattered throughout those specific project pages outside the main docs.  

I figured a "Caveats" section was appropriate since Docker, VirtualBox, & Vagrant weren't hard dependencies in the formula but are required to use the included drivers.  Additionally, we couldn't include the community.docker Ansible collection as a formula dependency but it is also a manually required setup step by the user to use the Docker driver.

Do you have a suggestion on how this should be best addressed?  I feel something needs to be mentioned in Caveats to alert the user that full use of the formula functionality requires additional steps.  I felt putting the list of dependencies in the Caveats makes it easiest for the user to identify requirements and quickly install them if needed.

carlocab replied:
This is best addressed in upstream documentation. If it's not easy to find there, let's help them clean it up.

Each formula needs to be maintainable by contributors who don't use the formula (and know only how to use Homebrew), and adding `caveats` like this gets in the way of that, as one needs to actually know how to use Molecule to understand what this says and whether it still applies.

SMillerDev replied:
As a rule, caveats are only to notify users of Homebrew inconsistencies compared to default upstream documentation.

carlocab reviewed (commented):
Thanks, @mpiekunk.

When you make further changes please include them in your existing commit, and amend your commit message to read
```
molecule: add docker driver
```
This is mentioned in the [commit style guide](https://docs.brew.sh/Formula-Cookbook):

>The preferred commit message format for simple version updates is `foobar 7.3` and for fixes is `foobar: fix flibble matrix.`.

mpiekunk commented:
I will change the commit message when addressing above comments.  Wasn't completely clear on what the appropriate format was since this technically isn't a fix but adding functionality.
